dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/cloudera/manager/ClouderaClustersTask.java

```diff
@@ -0,0 +1,124 @@
+ */
+package com.google.edwmigration.dumper.application.dumper.connector.cloudera.manager;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.common.collect.ImmutableList;
+import com.google.common.io.ByteSink;
+import com.google.edwmigration.dumper.application.dumper.connector.cloudera.manager.ClouderaManagerHandle.ClouderaClusterDTO;
+import com.google.edwmigration.dumper.application.dumper.connector.cloudera.manager.dto.ApiClusterDto;
+import com.google.edwmigration.dumper.application.dumper.connector.cloudera.manager.dto.ApiClusterListDto;
+import com.google.edwmigration.dumper.application.dumper.task.TaskRunContext;
+import java.io.Writer;
+import java.net.URI;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+import org.apache.http.HttpStatus;
+import org.apache.http.client.methods.CloseableHttpResponse;
+import org.apache.http.client.methods.HttpGet;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.util.EntityUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ClouderaClustersTask extends AbstractClouderaManagerTask {
+
+  private static final Logger LOG = LoggerFactory.getLogger(ClouderaClustersTask.class);
+
+  private final ObjectMapper objectMapper = new ObjectMapper();
+
+  public ClouderaClustersTask() {
+    super("clusters.json");
+  }
+
+  @CheckForNull
+  @Override
+  protected Void doRun(
+      TaskRunContext context, @Nonnull ByteSink sink, @Nonnull ClouderaManagerHandle handle)
+      throws Exception {
+    CloseableHttpClient httpClient = handle.getHttpClient();
+
+    ApiClusterListDto clusterList;
+
+    if (context.getArguments().getCluster() != null) {
+      final String clusterName = context.getArguments().getCluster();
+      try (CloseableHttpResponse clusterResponse =
+          httpClient.execute(new HttpGet(handle.getApiURI() + "/clusters/" + clusterName))) {
+
+        ApiClusterDto cluster =
+            objectMapper.readValue(
+                EntityUtils.toString(clusterResponse.getEntity()), ApiClusterDto.class);
+
+        clusterList = new ApiClusterListDto();
+        clusterList.setItems(ImmutableList.of(cluster));
+      }
+    } else {
+      LOG.info("'--cluster' argument wasn't provided. Collect all available clusters.");
+
+      try (CloseableHttpResponse clustersResponse =
+          httpClient.execute(new HttpGet(handle.getApiURI() + "/clusters"))) {
+        String clustersJson = EntityUtils.toString(clustersResponse.getEntity());
+
+        clusterList = objectMapper.readValue(clustersJson, ApiClusterListDto.class);
+      }
+    }
+
+    try (Writer writer = sink.asCharSink(StandardCharsets.UTF_8).openBufferedStream()) {
+      writer.write(objectMapper.writeValueAsString(clusterList));
```

Comment:
This seems to only write `ApiClusterListDto@abc123` rather than the data in this dto.

My suggestion is to log the items directly without wrapping.

If you will change this behavior, please also include it in unit tests.

Reply:
No, it's write json with the structure of the class, such behavior is the motivation to introduce `ApiClusterListDto`. 

It's also covered by unit test `com.google.edwmigration.dumper.application.dumper.connector.cloudera.manager.ClouderaClustersTaskTest:142`

Reply:
True, I forgot that jackson can convert this to string by itself with reflection. I've tested this, the output looks like: `{"items": [ obj#1, obj#2, ... ]}` where `obj#n` are items as objects

Still, I'd suggest to simplify this format and write just the list without the single-entry dictionary - `[ obj#1, obj#2, ... ]`. To me it looks like we won't need the wrapping ListDto.

For the tests, I meant comparing this to a fixed expected value. The test you've mentioned checks if we're able convert both ways without errors, but adding a value assertion would be helpful to see what we're converting to.